### DISCUSSION TITLE
Set custom initialization for record members

### DISCUSSION THREAD

**Discussion by stefanos82** [edubart/nelua-lang#77] [Q&A] [answered]
Why can't we set the default values ourselves upon need?

```lua
local Person = @record{
    name: string = "stefanos82",
    age: integer
}

local person: auto = Person{}
print('Name:', person.name) -- should print Name: stefanos82
print('Age:', person.age) -- should print 0
```

**Comment by stefanos82** [comment]
Ah because we can do it like this
```lua
-- typed initialization
local a: Person = {name = "Mark", age = 20}
print(a.name, a.age)
```
So, is `record` behaving like Java's `interface` in a way?

**Comment by edubart** [comment, marked as answer]
```lua
local Person = @record{
    name: string = "stefanos82",
    age: integer
}
```

This is a type declaration, not a runtime variable, and types in Nelua cannot have default initialization, this is a design decision. To initialize a variable with custom values you should use a function or initialize like you mentioned later:

```lua
local a: Person = {name = "Mark", age = 20}
```

I suggest to add `: type` in type declarations to be more explicit that it's a type declaration and not a runtime variable declaration, thus avoiding confusion:

```lua
local Person: type = @record{
    name: string = "stefanos82",
    age: integer
}
```

> So, is record behaving like Java's interface in a way?

No, a `record` is more like `struct` in C++, they are used to store a fixed data structure, but can also have methods. Similar to tables in Lua, where you use tables to store a dynamic structure and tables can also have methods (via metatables). Unlikely Java interfaces, by default they cannot be used to extend methods or fields of other records, although this could be done with some meta programming, something similar to this is done as an example in [record inheretance example](https://github.com/edubart/nelua-lang/blob/master/examples/record_inheretance.nelua).

**Reply by stefanos82** [reply]
> No, a `record` is more like `struct` in C++

Yeah, I was just reading some C code and realized it's more like a `typedef`-ed `struct`.

I thought `type` was supposed to be used behind the scenes in metaprogramming with preprocessor, that's why I didn't touch it.
